### PR TITLE
Webhook

### PR DESCRIPTION
**Pull request by mikemilano**
This example forwards workflow data to a url defined in the config file. An api_key may also be defined in the config for authentication on the receiving end.

There is also a d7 module provided which exposes a menu (`quicksilver/:api_key`). It forwards the received data to `hook_quicksilver($data)`.

An API key will automatically be generated and a sample `webhook.json` file is provided based on actual environment data.

Test mode can be turned on to print the received data to the db log.

### CODE REVIEW

**Comment by mikemilano** [comment]
I'm going to move the webhook.json data into the main plugin. The data isn't that sensitive and since it forwards requests across all systems it would be a pain to manage the 3 different json files.

**Comment by greg-1-anderson** [comment]
Cloudflare_cache example included here by mistake.

File: webhook/d7_module/quicksilver/quicksilver.module [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+function quicksilver_webhook_handler() {
+  $success = FALSE;
+  $message = '';
+  $submitted_key = arg(1);
+  $api_key = variable_get('quicksilver_api_key', '');
+
+  if (isset($submitted_key) && $submitted_key == $api_key) {
+    $success = TRUE;
+    $data = array(
```
Comment (greg-1-anderson) [inline webhook/d7_module/quicksilver/quicksilver.module]: Maybe $data should just be $_POST.  The module will be able to read $_ENV directly.  I don't think that `$_POST['payload']` is correct.

File: webhook/webhook.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// Retrieve webhook config data
+$config = json_decode(file_get_contents($_SERVER['HOME'] . '/files/private/webhook.json'), 1);
+if ($config == FALSE) {
+  die('files/private/webhook.json found. Aborting!');
+}
+
+send_data($config);
+
+function send_data($config) {
+  $payload = http_build_query(array('payload' => $_POST, 'env' => $_ENV));
```
Comment (greg-1-anderson) [inline webhook/webhook.php]: Maybe we should just send $_POST here. Module can read $_ENV directly.

**Comment by greg-1-anderson** [comment]
Looks good.  Have not tested, but +1 on code and docs.  Recommend removing $_ENV from the hook payload.

**Comment by mikemilano** [comment]
@greg-1-anderson I think at the `PANTHEON_SITE_NAME` is the main value that might be used for external processing. What do you think about just stuffing that in the payload as `site_name`? I think it would be fine to remove $_ENV then.

**Comment by greg-1-anderson** [comment]
The module should be able to access `$_ENV['PANTHEON_SITE_NAME']` or `getenv("PANTHEON_SITE_NAME")` directly.

**Comment by mikemilano** [comment]
The use case is that I have a separate application receiving the webhook (maybe drupal, maybe not) handling events for multiple sites.

**Comment by greg-1-anderson** [comment]
Ah.  Good point.  If the target is off-Pantheon, then these values might be relevant.

**Comment by greg-1-anderson** [comment]
I'd recommend taking just PANTHEON_SITE_NAME and PANTHEON_ENVIRONMENT, and including them somewhere in your payload -- exact structure not too important.  The other items in $_ENV are mostly not useful, and a couple (hash salt, db password) would actually be a security concern to include in an off-platform web request.

**Comment by mikemilano** [comment]
@greg-1-anderson I believe the last 2 commits addresses the concerns.

File: webhook/webhook.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
+// TODO: this header should not be required...
+header('Content-Type: text/plain; charset=UTF-8');
+
+// Retrieve webhook config data
+$config = json_decode(file_get_contents($_SERVER['HOME'] . '/files/private/webhook.json'), 1);
```
Comment (greg-1-anderson) [inline webhook/webhook.php]: Could still use better error checking here.

File: webhook/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+### Example POST data sent to the webhook url ###
+
+Below is an example of the data sent as a POST request to the `url` defined in `webhook.json`.
+
+This is also the same `$data` passed to `hook_quicksilver($data)` in the drupal module.
+
+```
+Array
```
Comment (greg-1-anderson) [inline webhook/README.md]: Update sample data below to match what the code is providing.

**Comment by mikemilano** [comment]
@greg-1-anderson Regarding the api key, do you think we should just remove the json file and set the endpoint url and api key inline? We have good examples of how to use secrets in other modules and it would make it less complicated to understand what's going on here.

**Comment by mikemilano** [comment]
Also, do you guys have a preference for sample name data in documentation? i.e. the values of the sample payload keys for first/last/full name.

**Comment by greg-1-anderson** [comment]
Yes, I think it would be a good idea to remove the "api key" and just hardcode a URL inline. In fact, this might be a good place to give an example of building a URL from the pantheon site name and environment variables, so it is config-free.  Add a comment to recommend changing the URL as needed, and provide a pointer to another example that manages secrets for users who need an API key.

As far as example values in the samples, I think it's fine to use your name, or John Doe, or a made-up person or a famous person, as the mood suits you. I don't think we have policies on this.  Just don't use real email addresses, or real database passwords.  :smile:

**Comment by mikemilano** [comment]
in terms of the url builder, again, the endpoint won't always be a pantheon site. another reason for the api key.. and the drupal module sample already has a simple api key generator in it.

**Comment by greg-1-anderson** [comment]
Yes, take care of non-Pantheon targets with a comment that advises the user to change the URL to suit.

I think the sample is clearer without the API hook, but I will defer to @joshkoenig on whether he wants you to remove that feature, as it's also a side benefit for those who might want to use the sample module as a starting point.

**Comment by joshkoenig** [comment]
Given how simple the JSON config file is, I'd argue to just put the config inline and leave it up to people who implement to do their own breakout as needed. 

The internet standard in webhookery is copy/pasta URLs. That's what GitHub does, as well as IFTTT, etc. Make it really simple.

I'd also recommend making the `send_data` function take a URL and a payload, which is pretty hacker friendly, rather than magically getting the `$_POST` inside, and taking a specific style of config that we invent.

```php
$url = 'https://maker.ifttt.com/trigger/{event}/with/key/mkjz03u5drFpr7amGuKP9CpdDbGwz36Rj1ilg08rjBk'
$payload = $_POST;
$payload['extra_key'] = 'something I needed to add';
send_data($url, $payload)
```

**Comment by mikemilano** [comment]
This should be ready for review. I just removed the drupal module as it adds potential confusion for an example that simply needs to show how to forward workflow data to a url. If it is wanted back, I can do the refactoring based on the latest changes here.

**Pull request closed without merging**